### PR TITLE
proc/gdbserial/gdbserver: Don't pass DYLD/LD to debugserver

### PR DESCRIPTION
**Pull request by aviramha**
This change fixes recent situations where you can't really load to debugserver/lldb since of new protections
but you still want to load to the debugged process delve already supports moving the parameters to the child using -e, this just supresses issues being loaded to the debugger itself

### CODE REVIEW

File: pkg/proc/gdbserial/gdbserver.go [hunk 1]
```diff
@@ -547,6 +547,14 @@ func LLDBLaunch(cmd []string, wd string, flags proc.LaunchFlags, debugInfoDirs [
 
 	if runtime.GOOS == "darwin" {
 		process.Env = proc.DisableAsyncPreemptEnv()
+		// filter out DYLD_* environment variables
```
Comment (aarzilli) [inline pkg/proc/gdbserial/gdbserver.go:550]: Explain why here.

**Review by aarzilli — COMMENTED** [review]
Can you be less vague about what problem this fixes? You say "recent situations" and "new protections" but what are those?

File: pkg/proc/gdbserial/gdbserver.go (hunk 1 above)
Comment (aviramha) [reply pkg/proc/gdbserial/gdbserver.go:550]: To have all the darwin env changes in the same spot - or you want me to add a comment saying that?

**Comment by aviramha** [comment]
> Can you be less vague about what problem this fixes? You say "recent situations" and "new protections" but what are those?

Yes, sorry. The changes aren't documented and I am not sure what they changed exactly but I'll try to do my best to explain here.
Until macOS Ventura, you could use `DYLD_INSERT_LIBRARIES` to load a custom dylib into `debugserver`. It would require you to build it with arm64e architecture (or fat binary that contains it as well).
Since macOS Ventura release, `debugserver` would crash saying no matching architecture is found, while all relevant architectures exist.
This fix let's us "skip" loading to `debugserver` when we actually want to load to the debugged process (that we were okay with loading to, because in that case we just did nothing)
There was already a feature to add `-e` to `debugserver` arguments of `DYLD_INSERT_LIBRARIES`/`LD_PRELOAD` so I assume there are more users that use this capability to load into child processes. The only issue is that environment variable being "leaked" to debugserver, which doesn't work anymore in macOS Ventura.
I'm currently chasing Apple engineers to understand if that's a desired change from their side (other binaries that they don't want custom dylibs loaded into have different approach - they just ignore the env var. In case of this binary it just crashes.)
Relevant issue in our project repository:
https://github.com/metalbear-co/mirrord/issues/677

**Comment by aarzilli** [comment]
> Until macOS Ventura, you could use `DYLD_INSERT_LIBRARIES` to load a custom dylib into `debugserver`

As far as I know this is not true, see:

https://stackoverflow.com/questions/33587130/why-doesnt-lldb-forward-my-environment-variable-anymore/33589760#33589760

macOS used to strip all DYLD_ environment variables before running protected processes. This is why getLdEnvVars exists.

File: pkg/proc/gdbserial/gdbserver.go (hunk 1 above)
Comment (aarzilli) [reply pkg/proc/gdbserial/gdbserver.go:550]: I think this comment should explain why we are filtering out DYLD_ variables.
Basically it should repeat what the commit message says and both should be more precise about what changed and when it did than what the current commit message is, if possible.

**Comment by aviramha** [comment]
> 

I was well aware of that - but specifically `debugserver` and other binaries weren't protected like other binaries - i.e you could use `DYLD_INSERT_LIBRARIES`. Post Ventura if you try to use it it crashes instead of stripping it like other protected binaries. That's why this patch exists.

File: pkg/proc/gdbserial/gdbserver.go (hunk 1 above)
Comment (aviramha) [reply pkg/proc/gdbserial/gdbserver.go:550]: Understood - I'll amend in few minutes.

Comment (aviramha) [reply pkg/proc/gdbserial/gdbserver.go]: I hope it is more clear now.

**Review by aarzilli — APPROVED** [review]
LGTM

**Comment by aviramha** [comment]
Thank you. Anything that I need to do?

**Comment by aarzilli** [comment]
No.

**Review by derekparker — APPROVED** [review]
LGTM

**Pull request merged**